utils/schemaValidation.js

```diff
@@ -0,0 +1,192 @@
+//
+// Returns an empty string (a falsy value) if the property is valid to the
+// schema. Otherwise, returns a non-empty string (a truthy value) specifying the
+// validation error message.
+//
+// Preconditions:
+// - Requires the property to be already in the schema.
+function getPropertyViolation(obj, property, schema) {
+  // if property is not in the schema, return an error string
+  // (this should never happen)
+  if (!hasProperty(schema, property)) {
+    return `Constraint violated: Property "${property}" not in schema\n\n`;
+  }
+
+  // if object does not have the property at all, return an error string
+  if (!hasProperty(obj, property)) {
+    return `Constraint violated: Property "${property}" not in input\n\n`;
+  }
+
+  // at this point, it is guaranteed that both object and schema have a matching
+  // property
+  let isValid;
+
+  // validate the property based on its type
+  switch (schema[property].type) {
+    case Type.integer:
+      isValid = validator.isInt(obj[property] + "", {
+        min: schema[property].minValue,
+        max: schema[property].maxValue
+      });
+      break;
+
+    case Type.string:
+      isValid = validator.isLength(obj[property] + "", {
+        min: schema[property].minLength,
+        max: schema[property].maxLength
+      });
+      break;
+
+    case Type.timestamp:
+      // time format must be ISO 8601, e.g. "2020-12-31T23:59:59Z"
+      isValid = validator.isISO8601(obj[property] + "", {
+        strict: true
+      });
+      break;
+
+    // if property is not of the allowed types, it's not valid
+    default:
+      isValid = false;
+      break;
+  }
+
+  // return an empty string if pass the validator or return a non-empty error
+  // string otherwise
+  return isValid ? "" : (schema[property].getErrorMessage() + "\n\n");
+}
+
+// Sanitizes an object using a provided schema by extracting valid properties to
+// a new object and returns that object.
+//
+// Note:
+// - This function assumes the object has been validated against the schema.
+// - This function does not support nested objects or with arrays inside.
+function sanitizeUsingSchema(obj, schema) {
+  const validObj = {};
+  if (obj) {
+    // since this function assumes the object has been validated against the
+    // schema, a property is valid if it is in the schema
+    Object.keys(schema).forEach(key => {
+      if (hasProperty(obj, key)) {
+        validObj[key] = obj[key];
+      }
+    });
+  }
+  return validObj;
+}
+exports.sanitizeUsingSchema = sanitizeUsingSchema;
+
+// Schema of an applied Plan used for the validator and the database.
+const planSchema = {
+  status: {
+    required: true,
+    type: Type.integer,
+    minValue: 0,
+    maxValue: 4,
+    getErrorMessage: function() {
+      return "Constraint violated: Invalid plan status\n" +
+        "Plan status must be 0 (Rejected), 1 (Awaiting Student Changes) " +
+        "2 (Awaiting Reivew), 3 (Awaiting Final Review), or 4 (Accepted).";
+    }
```

Comment:
I may not be understanding this correctly, but a status should not be required when the plan is submitted. The plan will always have a default status, which then can be updated by advisors at a later time.

Reply:
I heard you this morning but forgot to update it :D. Thanks!